### ISSUE TITLE
RKE2 Restore Hangs "bootstrap data is already found and encrypted with a different token"

### ISSUE THREAD

**Issue by braunsonm** [rancher/rancher#36298]
**Rancher Server Setup**
- Rancher version: 2.6.4
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): Helm
  
**Information about the Cluster**
- Kubernetes version: 1.21.5
   
**Describe the bug**
When attempting to restore RKE2 into another cluster, the process hangs with "bootstrap data is already found and encrypted with a different token".

**To Reproduce**
1. Create an RKE2 cluster
2. Back it up
3. Delete it
4. Create another RKE2 cluster
5. Try to restore it with the previous backup

**Result**

**Expected Result**
Backup should succeed

**Comment by braunsonm** [comment]
Confirmed with RKE2 folk that this is a feature Rancher would need to support since the node token has to be provided in order to decrypt: https://github.com/rancher/rke2/issues/2410

**Comment by Oats87** [comment]
Theoretically, with a little cluster manipulation, this should be possible provided you know what the server/agent token of the previous cluster is. 

https://github.com/rancher/rancher/blob/c4d59fa88e2313e1446ba542a21c39e15a17f375/pkg/provisioningv2/rke2/planner/planner.go#L1029-L1075

We create a secret using the cluster name + `-rke-state` that contains the server/agent tokens. If this file is created before the cluster is actually created, it should inherit this file. Then, when the cluster is created and restored from snapshot, the token will line up and the bootstrap data will be able to be decryptd.

For example, the file looks like: 

```
> kubectl get secret -n fleet-default tester-rke-state -o json
{
    "apiVersion": "v1",
    "data": {
        "agentToken": "em4ycjlkanNyZDZxY2R6Yjd0bng2Y<LOL>Iyc2RncG42cXNzczlyeGNrd3poNDR0N2xzZDR2",
        "serverToken": "dnhibHJ3cWg1dDZ4ZnIydGp3djhxc<LOL>Y1NzdndGxoZnA4OGpibms0MmtndG5tYmxkNnNu"
    },
    "kind": "Secret",
    "metadata": {
        "creationTimestamp": "2022-05-12T21:01:58Z",
        "name": "tester-rke-state",
        "namespace": "fleet-default",
        "ownerReferences": [
            {
                "apiVersion": "rke.cattle.io/v1",
                "kind": "RKEControlPlane",
                "name": "tester",
                "uid": "70fecb17-804d-47a4-a614-9291df0ce5b8"
            }
        ],
        "resourceVersion": "37232410",
        "uid": "f24a9758-5512-4ad8-ad44-e5e8b5df0fe0"
    },
    "type": "rke.cattle.io/cluster-state"
}
```

**Comment by braunsonm** [comment]
@Oats87 Thanks so much, it is great to know that there is at least a work around.

**Comment by sass1997** [comment]
@Oats87 should this workaround work for 2.6.7. I just did following steps.
- Deploy rke2 cluster
- register nodes
- deploy some example resources
- create a snashot
- export <cluster-name>-rke-state secret
- delete cluster and nodes
- create secret on base of exported
- create cluster again
- secret was overwritten after creating the cluster

**Comment by khushboo-rancher** [comment]
I tried the workaround provided by @Oats87 [here](https://github.com/rancher/rancher/issues/36298#issuecomment-1125526213), looks like if we create secret `cluster name + -rke-state` before the cluster creation, Rancher keeps deleting it.
I also tried replacing the secret which was created by Rancher while provisioning the cluster, but no luck.

**Comment by Oats87** [comment]
@khushboo-rancher Please show the corresponding logs (and the secret you attempted to create, you can redact the token data if you want)

I am able to create as many `-rke-state` secrets as necessary, and they are not deleted. I don't know why your secret would be deleted, unless you are not following the procedure.

**Comment by voarsh2** [comment]
> * 
> * create secret on base of exported
> * create cluster again
> * secret was overwritten after creating the cluster

Not sure what this means.
If I delete a cluster and create a new one in RKE 2 (via Rancher) I can't create secrets before the deployment. (cluster creation)...?

**Comment by voarsh2** [comment]
> I tried the workaround provided by @Oats87 [here](https://github.com/rancher/rancher/issues/36298#issuecomment-1125526213), looks like if we create secret `cluster name + -rke-state` before the cluster creation, Rancher keeps deleting it. I also tried replacing the secret which was created by Rancher while provisioning the cluster, but no luck.

You're right. I am trying on 2.7.5 - I put the cluster_name>-rke-state secret in the local cluster (Rancher) and it shortly gets deleted.......

**Comment by Oats87** [comment]
Your secret type cannot be opaque, it must be `rke.cattle.io/cluster-state`.

**Comment by hwo-wd** [comment]
I might be a bit late to chime in, but the [outlined procedure](https://github.com/rancher/rancher/issues/36298#issuecomment-1125526213) works nice on Rancher v2.11.x (tested a single-node cluster so far), and I've identified the confusion about the auto-deleting secret: **prior to creating** the secret, one needs to **remove the `.metdata.ownerReferences`** as this points to the previous cluster (which is by now not existent anymore and Rancher will auto-prune it shortly after creation otherwise).
Anyway, thanks @Oats87!
